pkg/apis/discovery/v1beta1/conversion.go

```diff
@@ -0,0 +1,239 @@
+)
+
+const (
+	v1beta1Topology     = "endpointslice.kubernetes.io/v1beta1-topology"
+	v1beta1ZoneTopology = "topology.kubernetes.io/zone"
+)
+
+type EndpointSliceTopology struct {
+	Strings  []string      `json:"s"`
+	Topology []map[int]int `json:"t"`
+}
+
+func Convert_v1beta1_EndpointSlice_To_discovery_EndpointSlice(in *v1beta1.EndpointSlice, out *discovery.EndpointSlice, s conversion.Scope) error {
+	// zone conversion is taken care by endpoint conversion
+	if err := autoConvert_v1beta1_EndpointSlice_To_discovery_EndpointSlice(in, out, s); err != nil {
+		return err
+	}
+
+	// Use maps initially to keep track of unique keys and counts of the values
+	keyMap := make(map[string]int)
+	valueCounts := make(map[string]int)
+	keysSize := 0
+	valuesSize := 0
+
+	// gather unique keys, values and value counts
+	for _, endpoint := range in.Endpoints {
+		for k, v := range endpoint.Topology {
+			// Don't include zone, as it will be covered by the zone field on the Endpoint
+			if k == v1beta1ZoneTopology {
+				continue
+			}
+
+			if _, ok := keyMap[k]; !ok {
+				keyMap[k] = 0
+				keysSize += len(k)
+			}
+
+			if _, ok := valueCounts[v]; !ok {
+				valueCounts[v] = 0
+				valuesSize += len(v)
+			}
+			valueCounts[v]++
+		}
+	}
+
+	approximateSize := approximateAnnotationSize(keyMap, valueCounts, len(in.Endpoints), keysSize, valuesSize)
+
+	maxSize := 256*(1<<10) - sizeOfAnnotations(out.Annotations)
+
+	removedStrings := 0
+	for approximateSize > maxSize {
+		keyToRemove := ""
+		for k := range keyMap {
+			if k != "kubernetes.io/hostname" && k != "topology.kubernetes.io/region" {
+				keyToRemove = k
+				break
+			}
+		}
+		// Either no more keys exist or remaining keys are standard keys
+		// No other way to preserve the data, so throw an error
+		if keyToRemove == "" {
+			return fmt.Errorf("unable to convert topology fields on v1beta1.EndpointSlice")
+		}
+		delete(keyMap, keyToRemove)
+		keysSize -= len(keyToRemove)
+		removedStrings++
+
+		for _, endpoint := range in.Endpoints {
+			val, ok := endpoint.Topology[keyToRemove]
+			if !ok {
+				continue
+			}
+			valueCounts[val]--
+			if valueCounts[val] == 0 {
+				delete(valueCounts, val)
+				valuesSize -= len(val)
+				removedStrings++
+			}
+		}
+
+		approximateSize = approximateAnnotationSize(keyMap, valueCounts, len(in.Endpoints), keysSize, valuesSize)
+	}
+
+	var strings []string
+	stringIndex := make(map[string]int)
+	var topologies []map[int]int
+	for _, endpoint := range in.Endpoints {
+		topology := make(map[int]int)
+		for k, v := range endpoint.Topology {
+			var kIndex, vIndex int
+			var ok bool
+			if kIndex, ok = stringIndex[k]; !ok {
+				if _, ok := keyMap[k]; !ok {
+					continue
+				}
+				kIndex = len(strings)
+				strings = append(strings, k)
+				stringIndex[k] = kIndex
+			}
+
+			if vIndex, ok = stringIndex[v]; !ok {
+				vIndex = len(strings)
+				strings = append(strings, v)
+			}
+			topology[kIndex] = vIndex
+		}
+		topologies = append(topologies, topology)
```

Comment:
This results in a v1 object that has a parallel slice of endpoints and topology annotations. Does that means if a v1 API client adds or removes an endpoint from the list, it is also responsible for rewriting the annotation to keep it in sync? Are we confident clients will be able to do that correctly? What are the implications if they do not?

Reply:
THis was the plan but a field is looking better and better